src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,177 @@
+import { CoinGeckoPrice } from "../coingecko";
+import { Logger } from "../relayFeeCalculator"; // @todo: Relocate Logger to utils?
+
+export { Logger }; // Permit adapters to import a common definition.
+export type TokenPrice = CoinGeckoPrice; // Temporary inversion; CoinGecko should source from here.
+
+// Represents valid source for spot prices (Across API, CoinGecko, ...)
+export interface PriceFeedAdapter {
+  readonly name: string;
+  getPriceByAddress(address: string, currency: string): Promise<TokenPrice>;
+  getPricesByAddress(addresses: string[], currency: string): Promise<TokenPrice[]>;
+}
+
+// It's convenient to map TokenPrice objects by their address, but consumers typically want an array
+// of TokenPrice objects, so the address must also be embedded within the TokenPrice object. To
+// avoid storing multiple copies of the same TokenPrice, always use the lower-case variant of the
+// address when performing a lookup into the PriceCache, and substitute the provided address into the
+// TokenPrice object when returning to the caller.
+export type PriceCache = {
+  [address: string]: TokenPrice;
+};
+
+export function msToS(ms: number): number {
+  return Math.floor(ms / 1000);
+}
+
+export class PriceClient implements PriceFeedAdapter {
+  public readonly name: string = "PriceClient";
+  private _maxPriceAge = 300; // seconds
+  protected prices: {
+    [currency: string]: PriceCache;
+  } = {};
+
+  constructor(protected logger: Logger, readonly priceFeeds: PriceFeedAdapter[]) {
+    assert(logger, "No logging instance supplied.");
+    assert(priceFeeds.length > 0, "No price feeds supplied.");
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  listPriceFeeds(): string[] {
+    return this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    const tokenPrices = await this.getPricesByAddress([address], currency);
+    return tokenPrices[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    assert(this.priceFeeds.length > 0, "No price feeds are registered.");
+    const priceCache = this.getPriceCache(currency);
+
+    // Determine whether *all* prices are current.
+    const now = msToS(Date.now());
+    const missed: { [address: string]: number } = {};
+    addresses.forEach((address: string) => {
+      const addr = address.toLowerCase();
+      // note: If priceCache[addr] is undefined, it is assigned here.
+      const tokenPrice = (priceCache[addr] = priceCache[addr] ?? ({ price: 0, timestamp: 0 } as TokenPrice));
+      const age = tokenPrice ? now - tokenPrice.timestamp : Number.MAX_SAFE_INTEGER;
```

Comment:
nit: I don't think this can ever be undefined, so you should be able to just do:
```ts
const age = now - tokenPrice.timestamp;
```

Reply:
True - fixed: https://github.com/across-protocol/sdk-v2/pull/98/commits/0138242162a90c2574b75b0eab27ee93118fc413